### PR TITLE
Replace Makefile boilerplate by shell scripts

### PR DESCRIPTION
**Pull request by return42**
## What does this PR do?

This is an alternative to PR #2529, replacing the implementation of make targets by bash scripts.

All build tasks are implemented in a bash script `./manage` and the `./Makefile` is just a wrapper for IDE and CI.

## Why is this change important?

The initial Makefile boilerplate was copied from https://github.com/return42/cdb-tools/tree/master/utils and was good to get things done but it is to abstract and complicated to potential contributors.  Implementing more straight forward shell scripts is more common to the community.  

## How to test this PR locally?

Run all the make targets / use `make help` and have a look at the modified documentation.

### CODE REVIEW

**Comment by return42** [comment]
~~It seems CI has some [trouble](https://github.com/searx/searx/pull/2594/checks?check_run_id=1955597485#step:5:562) with the command [`git name-rev --name-only HEAD`](https://github.com/searx/searx/pull/2594/files#diff-181229424893bb65d94a74c2132b8b9e5adfe851464fdb5cb9f49e8a8204be7bR88) .. I will have a look at .. but the rest of this PR is ready to comment ...~~

update: I screwed back to the existing method using JamesIves/github-pages-deploy-actio, see https://github.com/return42/searx/blob/manage-script/.github/workflows/integration.yml#L75-L85

File: manage
Comment (return42) [inline manage]: On branch master, this should return `master` but the CI returns `remotes/pull/2594/merge` which might be due to the PR cycle of github. I will look for an alternative.

update: may be this will return 'master' when it gets merged? Compare https://github.com/return42/searx/runs/1955388092#step:5:563

Comment (dalf) [reply manage]: According to 
https://github.com/JamesIves/github-pages-deploy-action/tree/3.7.1#optional-choices

there is `SINGLE_COMMIT` option.

https://github.com/searx/searx/blob/5f4a085fc44ed64499d1aa5c2e29d2f861037b5f/.github/workflows/integration.yml#L75-L84

becomes

```yaml
    - name: Build documentation
      run: SEARX_DEBUG=1 make V=1 doc.html
    - name: Deploy
      if:  github.ref == 'refs/heads/master'
      uses: JamesIves/github-pages-deploy-action@3.7.1
      with:
        GITHUB_TOKEN: ${{ github.token }}
        BRANCH: gh-pages
        FOLDER: build/gh-pages
        CLEAN: true # Automatically remove deleted files from the deploy branch
        SINGLE_COMMIT: true
```

* On "Build documentation", the documentation is built. This always happens, on the master branch or in the PR.
* On "Deploy", the documentation is pushed to the gh-pages branch using `JamesIves/github-pages-deploy-action`. Only in the master branch.

Comment (return42) [reply manage]: OK, what do you suggest .. should we use `JamesIves/github-pages-deploy-action` further?

BTW: I forgot to remove https://github.com/searx/searx/blob/5f4a085fc44ed64499d1aa5c2e29d2f861037b5f/.github/workflows/integration.yml#L77-L84

In general I am not a fan boy of (Github-) actions I can't reproduce locally, that was also one reason for me to change back to a (more simple) script which can be executed locally. 

Another reason for me is, that I know git and bash but Github-Actions are new to me / and to be true I don't want to investigate to much time in tools that can only be used on a proprietary platform.

Another point that disturbs me with, is the the fact that this github-Action needs hundred of lines of code for a simple commit.

But you are the maintainer and you should decide what is the best for you .. 

What do you think, should screw it back to the old, existing method where the scripts compile the docs and _Ives_ deploys? 

I think we can discuss my doubt in a separated issue ..

Comment (return42) [reply manage]: FYI: I am working on a _rebase_ fixing this issue and some other issues I found.

**Comment by return42** [comment]
@dalf sorry for the noise / I screwed back to the existing method using JamesIves/github-pages-deploy-action and fixed some other minor issues (missing SPDX tags)

**Review by dalf — CHANGES_REQUESTED** [review]
Here some comments after reading the code.

The big question is where to declare the dependencies between the targets.

Most of them are declared in `Makefile`, but there are also sometimes declare in `manage`.

**Comment by return42** [comment]
> The big question is where to declare the dependencies between the targets.

I did not know this either .. but after reading your comments .. I think it is better to declare them all in the bash file

Your comments have helped me get a clearer view. With I think I will find better solutions and polish up the implementations .. not today .. tomorrow :-)

File: Makefile
Comment (return42) [reply Makefile]: I have a doubt .. for me, `clean` is like `run`, `install` and `uninstall` ... these targets are out  of the scope of the `./manage` script I, think.

If you think `clean` is better implemented in manage script I can do it  ... meanings?

File: manage
Comment (return42) [reply manage]: > There is another implementation line 659 of utils/lib.sh

That's right. For me `./lib` is _abstract_ implementing basic stuff. `./lib.sh` is sourced by a _script_ and the functions from can be overwritten in the _script_ to fit the needs of the _application_ the script serves.

Here the _assert_ function calls additional `buildenv` compared to the _assert_ from lib.  This _assert_ also lets die the script if there are any problems when building (e.g.) the buildenv.

I know that you like to have a _straight forward_ implementation but a t this point I would say a minimal _abstraction_ is helpful.  The `./lib.sh` I also use in other projects (I maintain this lib since years) .. take it like any other lib we have in our project (e.g. github-pages-deploy-action in the build or jquery in the client).

Hope it is OK for you.

File: utils/lib.sh
Comment (return42) [reply utils/lib.sh:543]: I can review the function to see if I can improve it but beside I think it is a lib that should work stand-alone (see https://github.com/searx/searx/pull/2594#discussion_r582774331) .. for me it is important that I can reuse this code in other projects .. having special implementation in every project's boilerplate will decrease the quality of all projects over the time.

File: utils/lib.sh
Comment (return42) [reply utils/lib.sh]: I don't know if I have you understand fully .. 

`pyenv.check` is a function that does nothing more the printing the python commands to check if the installation is OK.  Theses commands are piped to a python interpreter.  E.g. :

```python
pyenv.check() {
    echo "import yaml"
}
```
is the minimalist check we have theses days (compare with ./manage). This function has to be implemented by the _script_, the check can't be a part of the ./lib.sh (see also https://github.com/searx/searx/pull/2594#discussion_r582774331)

File: manage
Comment (return42) [reply manage:10]: `./config.sh` is used by all bash scripts (compare searx.sh, morty,sh ..) .. I think all processes should have the same environment, this is one of the big benefits of moving implementations from Makefile into a bash script I think.  

But you are right, at the moment there is no explicit  usage .. but if one wants, he is able to customize what ever environment he wants.

File: Makefile
Comment (return42) [reply Makefile]: > These targets are not exclusively design for the CI.

Even this targets are named ci, they are not dedicated only to ci .. ci. here means _compile all themes_  

But it might be better to name them themes.all ... and so on .. to be more precisely.

> Also to rename: ... to data.xyz

make  sense to me. I will do it.

File: Makefile
Comment (return42) [reply Makefile]: I think so, we have used terms like `docs-live` and `docs-clean`all over in the issue tracker and docs.  I will fix the docs, but for a period we should also prompt some _deprecated_ messages.

**Comment by return42** [comment]
`./Makefile` now wraps all commands from `./manage`. The `./Makefile` is now very simple and plain : 

- https://github.com/return42/searx/blob/manage-script/Makefile#L13-L23
- https://github.com/return42/searx/blob/manage-script/Makefile#L70-L88

A preview of the new Makefile documentation is here: https://return42.github.io/searx/dev/makefile.html

File: utils/lib.sh
Comment (return42) [reply utils/lib.sh:710]: The new implementation assure a _pyenv_ with searx installed by function `pyenv.install`. The `docs.html` and `docs.live` prerequisites reduce to:

    pyenv.install
    docs.prebuild

File: Makefile
Comment (return42) [reply Makefile]: Now it is `ci.test` and it is used in the CI / the first test is `test.pep8`

    ci.test: test.pep8 test.pylint test.unit test.robot

The current master does not run a `test.sh` / same now with `ci.test` / we can add `test.sh` in a different PR

File: Makefile
Comment (return42) [reply Makefile]: The `Makefile` implementation is now very clean a straight forward:

https://github.com/return42/searx/blob/manage-script/Makefile#L70-L88

File: Makefile
Comment (return42) [reply Makefile]: Now we have one block with _deprecation_ messages:

- https://github.com/return42/searx/blob/manage-script/Makefile#L90-L107

We can purge this block in later in a further release (but not in the next).

File: manage
Comment (return42) [reply manage]: The `pyenv.assert` function is no longer needed in the new implementation.

File: Makefile
Comment (return42) [reply Makefile]: The new implementation is more explicit, using [`MANAGE += docker.build docker.push`](https://github.com/return42/searx/blob/manage-script/Makefile#L76)

File: Makefile
Comment (return42) [reply Makefile]: Same here: the new implementation of `./manage` is more explicit: https://github.com/return42/searx/blob/manage-script/manage#L58-L60

File: Makefile
Comment (return42) [reply Makefile]: Now it is wrapped 1:1

- https://github.com/return42/searx/blob/manage-script/Makefile#L77- https://github.com/return42/searx/blob/manage-script/Makefile#L87-L88

File: Makefile
Comment (return42) [reply Makefile]: Same here: is now 1:1

File: Makefile
Comment (return42) [reply Makefile]: fixed: pep8 test runs before pylint test.

File: Makefile
Comment (return42) [reply Makefile]: Same here; commands are wrapped 1:1 and targets like `doc` has been moved to _deprecated_  section.

File: Makefile
Comment (return42) [reply Makefile]: fixed: have a look at the new [name schema](https://return42.github.io/searx/dev/makefile.html) , there are only some [make-targets left](https://github.com/return42/searx/blob/manage-script/Makefile#L16-L23)

File: utils/lib.sh
Comment (return42) [reply utils/lib.sh]: Now we have a less generic one in the manage script https://github.com/return42/searx/blob/manage-script/manage#L342-L357 

Since `pyenv.install` is a elementary prerequisite the script will die if the installation fails in any way (installation can not be assured).

File: utils/lib.sh
Comment (return42) [reply utils/lib.sh]: Now we have a less generic one in the manage script: https://github.com/return42/searx/blob/manage-script/manage#L359-L362

File: manage
Comment (return42) [reply manage]: > Some of the target dependencies are written in the Makefile.

All needed dependencies are now implemented in the `./manage` script. There are only rare exceptions for typical make-targets like `test.sh`, `clean`, `run` ... and `search.checker.%`.

File: manage
Comment (return42) [reply manage]: fixed / thanks!

**Comment by return42** [comment]
I rebuild to solve conflicts ... all requests from approving this PR has been treated .. and I finished my tests.

**Comment by return42** [comment]
I want to add some thoughts of mine about the longer journey of the build process and all this stuff what is not searx core and why I think this PR is important for.

The searx repository should focus on the core development of the searx server and get rid of all the instance build scripts, LXC stuff and so on.  My plan is to do this in separated steps / I can thing about these steps: 

1. replace makefiles by bash scripts (this PR)
2. chunk the lib.sh beast in different parts / those needed for build / those needed for developer tasks (`./manage`)
3. create a repository "searx-build" / move all the build (and LXC) stuff from searx core to the "searx-build" / future: prepare packages for the package mangers of the common linux distros in "searx-build".
4. Split the user-doc and ship with the instance  / move admin-doc to "searx-build".

I haven't thought all to its end and it will be a longer journey .. but I could imagine something like that.

@asciimoo what do you think, does this meet your interests, is it a way you would go with?

**Comment by asciimoo** [comment]
> @asciimoo what do you think, does this meet your interests, is it a way you would go with?

Yeah, looks good to me at first glance. Perhaps in the end we can get back the good old manage.sh what was the original "build" script before makefiles. =)

**Comment by asciimoo** [comment]
@return42 could you restore this branch?

**Review by asciimoo — APPROVED** [review]
Great, thanks! Let's proceed with the current state and we can iterate later according to the feedback of the community.

**Pull request merged**
